app/models/transformation_mapping.rb

```diff
@@ -1,9 +1,115 @@
 
   def destination(source)
     transformation_mapping_items.find_by(:source => source).try(:destination)
   end
+
+  # vm_list: collection of hashes, each descriping a VM.
+  def validate_vms(vm_list = nil)
+    vm_list.present? ? identify_vms(vm_list) : select_vms
+  end
+
+  private
+
+  def select_vms
+    valid_list = []
+
+    transformation_mapping_items.where(:source_type => EmsCluster).collect(&:source).each do |cluster|
+      cluster.vms.each do |vm|
+        reason = validate_vm(vm)
+        valid_list << describe_vm(vm, reason) if reason == VM_VALID
+      end
+    end
+
+    {"valid_vms" => valid_list}
+  end
+
+  def identify_vms(vm_list)
+    valid_list = []
+    invalid_list = []
+    conflict_list = []
+
+    vm_list.each do |row|
+      vm_name = row['name']
+
+      query = Vm.where(:name => vm_name)
+      query = query.where(:uid_ems => row['uid_ems']) if row['uid_ems'].present?
+      query = query.joins(:host).where(:hosts => {:name => row['host']}) if row['host'].present?
+      query = query.joins(:ext_management_system).where(:ext_management_systems => {:name => row['provider']}) if row['provider'].present?
+
+      vms = query.to_a
+      invalid_list << describe_non_vm(vm_name) if vms.empty?
+
+      if vms.size == 1
+        reason = validate_vm(vms.first)
+        (reason == VM_VALID ? valid_list : invalid_list) << describe_vm(vms.first, reason)
+      else
+        vms.each { |vm| conflict_list << describe_vm(vm, VM_CONFLICT) }
+      end
+    end
+
+    {
+      "valid_vms"    => valid_list,
+      "invalid_vms"  => invalid_list,
+      "conflict_vms" => conflict_list
+    }
+  end
+
+  def describe_non_vm(vm_name)
+    {
+      "name"   => vm_name,
+      "reason" => VM_NON_EXIST
+    }
+  end
+
+  def describe_vm(vm, reason)
+    # formate to be finalized
+    {
+      "name"           => vm.name,
+      "cluster"        => vm.ems_cluster.name,
+      "path"           => "#{vm.ext_management_system.name}/#{vm.parent_blue_folder_path(:exclude_non_display_folders => true)}",
+      "allocated_size" => vm.allocated_disk_storage,
+      "id"             => vm.id,
+      "reason"         => reason
+    }
+  end
+
+  def validate_vm(vm)
+    # a valid vm must find all resources in the mapping and has never been migrated
+    invalid_list = []
+    invalid_list << N_("cluster: %{name}") % {:name => vm.ems_cluster.name} unless valid_cluster?(vm)
+
+    invalid_storages = validate_storages(vm)
+    invalid_list << N_("storages: %{list}") % {:list => invalid_storages.collect(&:name).join(", ")} if invalid_storages.present?
+
+    invalid_lans = validate_lans(vm)
+    invalid_list << N_("lans: %{list}") % {:list => invalid_lans.collect(&:name).join(', ')} if invalid_lans.present?
+
+    return N_("Not defined source for this migration - %{list}") % {:list => invalid_list.join('. ')} if invalid_list.present?
+
+    vm.valid_for_v2v_migration? ? VM_VALID : N_('VM is not valid for migration')
```

Comment:
I still think we should give a specific reason instead of `VM is not valid for migration`. Above you already collect detail reason for invalid cluster, storages, and lans. Here we would better say `VM has been migrated` or `VM has been added to other plan`.
In order to do this, consider change method `valid_for_v2v_migration?` (boolean) to `validate_for_v2v_migration` (string)